### DISCUSSION TITLE
SNMP support in Zephyr

### DISCUSSION THREAD

**Discussion by cbrake** [zephyrproject-rtos/zephyr#80648] [General]
I'm looking at adding SNMP support to Zephyr. So far I've looked at various Arduino implementations, but they are all C++, and I'd like to avoid that.

Porting pieces of net-snmp seems to be the best approach I've found yet, so working on that.

I don't have anything to share yet, but if anyone has interest, let me know.

**Comment by mariopaja** [comment]
@cbrake this is a topic that could be interesting for one of our projects. We have kind of skipped it because it was not a hard requirement for the endpoints, but definitely interesting to have it or being able to implement it

**Comment by jukkar** [comment]
I was looking at this some time ago, but did not had time to work on it.
I think the lwip snmp implementation might be easy to port to zephyr https://github.com/lwip-tcpip/lwip/tree/master/src/apps/snmp

**Reply by cbrake** [reply]
Thanks for the link -- this does look like a good starting point -- much better than anything I've found yet.

**Comment by ghost** [comment]
We are interested in this.

**Comment by ghost** [comment]
See also: https://github.com/zephyrproject-rtos/zephyr/issues/2329

**Comment by cbrake** [comment]
We are starting development in this external module:

https://github.com/simpleiot/zephyr-snmp

@htibosch is currently working on this

Looks like some cleanup will be required to upstream it (converting LWIP stuff to equivalent Zephyr), but hopefully we'll have something usable soon.

**Comment by cbrake** [comment]
@jukkar with SNMP, we are focusing on:

1. Traps (get sent out)
2. GetRequests (respond to requests)

For #2, would you advise creating an API like the HTTP server with callbacks that fire when requests come in, or is there a better way to do this?

**Reply by ghost** [reply]
Have you looked at https://docs.zephyrproject.org/latest/connectivity/networking/api/socket_service.html ?
Could be interesting for system level socket-services like this (see DHCP).

**Reply by cbrake** [reply]
thanks @clamattia that looks very useful!

**Comment by dowhile98** [comment]
I’ve been working on a project related to SNMP and have been testing it recently. I know this issue is quite old, but I’ll leave it here in case it’s still useful to you or you’d like to take a look at it.
https://github.com/dowhile98/zephyr-snmp

**Reply by jukkar** [reply]
@dowhile98 thanks for the link, the code looks very reasonable. Have you thought about trying to submit it to upstream zephyr?

**Comment by htibosch** [comment]
Thanks for remembering me. That is indeed a while ago.
But I'm back in my earlier job: development and maintenance of [FreeRTOS+TCP](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP), mostly working on IPv6.
Cheers
